### PR TITLE
fix: set default on reset of The Tab widget

### PR DESCRIPTION
**Pull request by techbhavin**
## Description

> Fixed to set a default value on reset action for the Tabs widget.

Fixes [#14694](https://github.com/appsmithorg/appsmith/issues/14694)

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

1. Reset the Tabs widget With the default value.
2. Reset the Tabs widget Without the default value.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by Rishabh-Rathod — COMMENTED** [review]
Could we please fix failing cypress test before the code review and also add cypress test case for verification of the fix.

File: app/client/src/widgets/TabsWidget/widget/index.tsx [hunk 1]
```diff
@@ -350,31 +357,36 @@ class TabsWidget extends BaseWidget<
     return [];
   };
 
+  setDefaultSelectedTabWidgetId = () => {
+    const visibleTabs = this.getVisibleTabs();
+    // Find the default Tab object
+    const selectedTab = find(visibleTabs, {
```
Comment (ashit-rath) [inline app/client/src/widgets/TabsWidget/widget/index.tsx:363]: Let's rename `selectedTab` to `defaultTab`?

File: app/client/src/widgets/TabsWidget/widget/index.tsx [hunk 2]
```diff
@@ -328,11 +328,18 @@ class TabsWidget extends BaseWidget<
...
+        if (Object.keys(this.props.tabsObj || {}).length) {
+          // If we have a defaultTab
+          if (this.props.defaultTab) {
+            this.setDefaultSelectedTabWidgetId();
+          } else {
+            // If no tab is selected
+            // Select the first tab in the tabs list.
+            this.props.updateWidgetMetaProperty(
+              "selectedTabWidgetId",
+              visibleTabs.length ? visibleTabs[0].widgetId : undefined,
```
Comment (ashit-rath) [inline app/client/src/widgets/TabsWidget/widget/index.tsx:340]: We could replace this with
```suggestion
              visibleTabs?.[0]?.widgetId,
```

File: app/client/src/widgets/TabsWidget/widget/index.tsx [hunk 3]
```diff
@@ -350,31 +357,36 @@ class TabsWidget extends BaseWidget<
...
+    // Find the default Tab object
+    const selectedTab = find(visibleTabs, {
+      label: this.props.defaultTab,
+    });
+    // Find the default Tab id
+    const selectedTabWidgetId = selectedTab
+      ? selectedTab.widgetId
+      : visibleTabs.length
+      ? visibleTabs[0].widgetId
+      : undefined; // in case the default tab is deleted
```
Comment (ashit-rath) [inline app/client/src/widgets/TabsWidget/widget/index.tsx:371]: We could also replace this with
```suggestion
    const selectedTabWidgetId = selectedTab?.widgetId ??  visibleTabs?.[0]?.widgetId;
```

Also please do modify `selectedTabWidgetId` to `defaultTabWidgetId`

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/Tab_reset_spec.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
+const LayoutPage = require("../../../../locators/Layout.json");
```
Comment (ashit-rath) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/Tab_reset_spec.js:1]: Observation:
1. The `it` block always represents one test case for a test spec. Right now it seems like each `it ` block here is a step leading to the last `it` block which is the actual test case to validate the reset functionality. So please add the whole test case in one it statement.
2. I think we should add another test case to check if the default tab is deleted and then the tab is reset, it should default to the first tab.

**Review by ashit-rath — COMMENTED** [review]
@techbhavin The `How Has This Been Tested?` section is for the devs to write down a summary of the tests that they have written; not "how to reproduce" the issue. It gives the QA/Reviewers/PMs/EMs a quick way to verify if all the cases are covered. 
Please do update this section.

File: app/client/src/widgets/TabsWidget/widget/index.tsx [hunk 5]
```diff
@@ -328,11 +328,18 @@ class TabsWidget extends BaseWidget<
...
-            visibleTabs[0].widgetId,
-          );
+        if (Object.keys(this.props.tabsObj || {}).length) {
+          // If we have a defaultTab
+          if (this.props.defaultTab) {
+            this.setDefaultSelectedTabWidgetId();
+          } else {
+            // If no tab is selected
+            // Select the first tab in the tabs list.
+            this.props.updateWidgetMetaProperty(
```
Comment (ashit-rath) [inline app/client/src/widgets/TabsWidget/widget/index.tsx:338]: Do we need the `else` case? Wouldn't the `setDefaultSelectedTabWidgetId` method deal with this case as well where the defaultTab is not present and assigns the `visibleTabs?.[0]?.widgetId`?

If yes, then we could remove the whole if-else and directly call the `setDefaultSelectedTabWidgetId`

File: app/client/src/widgets/TabsWidget/widget/index.tsx [hunk 6]
```diff
@@ -350,31 +357,34 @@ class TabsWidget extends BaseWidget<
...
+        "selectedTabWidgetId",
+        defaultTabWidgetId,
+      );
+    }
+  };
+
   componentDidMount() {
     const visibleTabs = this.getVisibleTabs();
     // If we have a defaultTab
     if (this.props.defaultTab && Object.keys(this.props.tabsObj || {}).length) {
```
Comment (ashit-rath) [inline app/client/src/widgets/TabsWidget/widget/index.tsx:386]: If we modify the whole `if-else` condition here to something like 
```js
if (Object.keys(this.props.tabsObj || {}).length) {
  this.setDefaultSelectedTabWidgetId();
}
```

Would it do the same that that we are doing here?

**Review by Rishabh-Rathod — COMMENTED** [review]
Please also remove deep equality check in `TabsWidget/widget/index.tsx`

`if (!isEqual(prevProps, this.props)) {`

As mentioned here - https://github.com/appsmithorg/appsmith/pull/13084#discussion_r926271194

cc: @ashit-rath

**Comment by ashit-rath** [comment]
@techbhavin I think we should relook at what we are doing in the `componentDidUpdate`. It seems like the logic is very similar to what we are doing in `setDefaultSelectedTabWidgetId`. 
Can you check and see if we can refactor and remove redundant bits?

File: app/client/src/widgets/TabsWidget/widget/index.tsx [hunk 7]
```diff
@@ -291,50 +291,33 @@ class TabsWidget extends BaseWidget<
...
-        const selectedTabWidgetId = selectedTab
-          ? selectedTab.widgetId
-          : undefined;
-        this.props.updateWidgetMetaProperty(
-          "selectedTabWidgetId",
-          selectedTabWidgetId,
-        );
-      }
-      // if selected tab is deleted
+    const visibleTabs = this.getVisibleTabs();
```
Comment (ashit-rath) [inline app/client/src/widgets/TabsWidget/widget/index.tsx:294]: The `componentDidUpdate` seems to check majorly 3 things
1. default tab change
2. a. selected tab changed and selectedId is not present in tabs
    b. selected tab changed and tabs not present
    c. selected tab changed and selectedTabId is null/undefined
3. selected tab but not present in visibleTabs.

They seems to be related to tab not being present in the tabs object expect point 1.

I feel like we can condense the whole componentDidUpdate with the following

```
const selectedTab = find(visibleTabs, {
  widgetId: this.props.selectedTabWidgetId,
});

if (this.props.defaultTab !== prevProps.defaultTab || !selectedTab) {
  this.setDefaultSelectedTabWidgetId();
}
```
Let me know if I'm missing something here or this doesn't cover the cases.

If this works fine then please do check if we have appropriate cypress test for all these scenarios.

**Review by ashit-rath — APPROVED** [review]
Great job @techbhavin

**Comment by chandannkumar** [comment]
Tested this PR and working as expected. Tested for
- Reset the Tabs widget With the default value.
- Reset the Tabs widget Without the default value.

**Review by ashit-rath — APPROVED** [review]

**Pull request merged**